[Lim Wei Jie] iP

lwj1711 opened the pull request:
# Duke

> “To accomplish a difficult task one must first make it easy.” – Martin Rubin

Duke helps you organize your tasks, using local storage to keep your files saved between sessions, making task management *easy* and *effortless* 

## Features
- [x] **Task Tracking**
Three different types of tasks:
  1. Todos: Tasks without a specific date
  2. Deadlines: Tasks to do **by** a specific date/time
  3. Events: Tasks to do **at** a specific date/time
- [x] **Mark Task As Done**
- [x] **Task Deletion**
- [x] **Find Tasks Matching Keywords**
- [x] **Print Current List of Tasks**
- [ ] **Reminders** (coming soon)


## Commands
- **todo**
- **deadline**
- **event**
- **list**
- **mark**
- **delete**
- **bye**

*Note that unrecognised commands will be met with a prompt to enter a recognised one instead*

## Extras
If you are a Java programmer, you can use it to practice Java too! Here's the `main` method:
```java
public class Duke {
    public static void main(String[] args) {
        new Duke("data/duke_list_data.txt").run();
    }
    ...
}
```
What are you waiting for! 🦄 
Download your Duke assistant now [here](https://github.com/lwj1711/ip)!

Commit:
Add Gradle support

lwj1711 committed:
Add assistant functionality

lwj1711 committed:
Add Task class
Update Duke functionality

lwj1711 committed:
Add mark as done feature.

lwj1711 committed:
Revert "Add mark as done feature."

This reverts commit 5d87069e82739024f68386ada7fd18166b725193.

lwj1711 committed:
Revert "Revert "Add mark as done feature.""

This reverts commit bf8f3579399c799031b52a7d7f28bf571a0b86a0.

lwj1711 committed:
Add ToDo, Deadline, Event classes and functionality

lwj1711 committed:
Implement testing

lwj1711 committed:
Implement error handling

lwj1711 committed:
Implement delete function

lwj1711 committed:
Add enums

lwj1711 committed:
Update for Java style conventions

lwj1711 committed:
Add Level-7 increments

lwj1711 committed:
Start Level-8 increments

lwj1711 committed:
Start Level-8 increments

lwj1711 committed:
Update Level-7 increments

lwj1711 committed:
Update Level-8 increments

lwj1711 committed:
Update .gitignore

lwj1711 committed:
Merge branch 'branch-Level-7'

lwj1711 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/InvalidCommandException.java
#	src/main/java/InvalidTaskException.java
#	src/main/java/MissingTaskException.java
#	src/main/java/MissingTimeException.java

lwj1711 committed:
Update Level-7 and Level-8 increments

lwj1711 committed:
Complete A-MoreOOP

lwj1711 committed:
Complete A-Packages

lwj1711 committed:
Complete A-JUnit

lwj1711 committed:
Complete A-Jar

lwj1711 committed:
Complete A-Jar

lwj1711 committed:
Complete A-JavaDoc

lwj1711 committed:
Complete A-CodingStandard

lwj1711 committed:
Complete Level-9

lwj1711 committed:
Merge branch 'branch-A-JavaDoc'

lwj1711 committed:
Merge branch 'branch-A-CodingStandard'

lwj1711 committed:
Merge branch 'branch-Level-9'

lwj1711 commented:
# Duke

> “To accomplish a difficult task one must first make it easy.” – Martin Rubin

Duke helps you organize your tasks, using local storage to keep your files saved between sessions, making task management *easy* and *effortless* 

## Features
- [x] **Task Tracking**
Three different types of tasks:
  1. Todos: Tasks without a specific date
  2. Deadlines: Tasks to do **by** a specific date/time
  3. Events: Tasks to do **at** a specific date/time
- [x] **Mark Task As Done**
- [x] **Task Deletion**
- [x] **Find Tasks Matching Keywords**
- [x] **Print Current List of Tasks**
- [ ] **Reminders** (coming soon)


## Commands
- **todo**
- **deadline**
- **event**
- **list**
- **mark**
- **delete**
- **bye**

*Note that unrecognised commands will be met with a prompt to enter a recognised one instead*

## Extras
If you are a Java programmer, you can use it to practice Java too! Here's the `main` method:
```java
public class Duke {
    public static void main(String[] args) {
        new Duke("data/duke_list_data.txt").run();
    }
    ...
}
```
What are you waiting for! 🦄 
Download your Duke assistant now [here](https://github.com/lwj1711/ip)!

lwj1711 committed:
Update Duke description

lwj1711 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

lwj1711 committed:
Complete A-Gradle

lwj1711 committed:
Merge branch 'branch-A-Gradle'

lwj1711 committed:
Complete A-CheckStyle

lwj1711 committed:
Merge branch 'branch-A-CheckStyle'

nguyiyang commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,279 @@
+    /**
+     * Marks the corresponding task as done and prints confirmation.
+     *
+     * @param toMark The index of the task to be marked.
```
Perhaps `toMark` can be renamed to something more intuitive like `index` here, same goes for `toDelete` in `deleteTask`.

nguyiyang commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,279 @@
+     * @throws MissingTimeException If time is unspecified after command.
+     */
+    public static void findTask(String input) throws MissingTaskException {
+        ArrayList<Task> toPrint = new ArrayList<>();
```
Should the `toPrint` variable have the word 'task' in its name? I feel like it would be clearer that the variable is referring to a `TaskList` that way.

nguyiyang commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,279 @@
+            return;
+        }
+
+        System.out.println("The matching "
+                + (toPrint.size() == 1 ? "task in your list is:" : "tasks in your list are:"));
```
I like how you considered the number of tasks and the string to print whether its singular or plural! :grinning:

lwj1711 committed:
Add GUI without Duke functionalities

nguyiyang commented on src/main/java/duke/command/Storage.java:
```diff
@@ -0,0 +1,126 @@
+     * @param tasks The array representing the task list.
+     */
+    public static void addTaskToList(String lineToAdd, ArrayList<Task> tasks) {
+        String taskDetails = lineToAdd.substring(7);
+
+        if (lineToAdd.charAt(1) == 'T') {
+            Task currentTask = new ToDo(taskDetails);
+            tasks.add(currentTask);
+            if (lineToAdd.charAt(4) == 'X') {
+                currentTask.markAsDone();
+            }
+        } else if (lineToAdd.charAt(1) == 'D') {
+            int separator = taskDetails.indexOf(" (by: ");
+            String taskName = taskDetails.substring(0, separator);
+            String timeFull = taskDetails.substring(separator + 6);
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
+            Task currentTask;
+
+            if (timeFull.length() > 12) {
+                String time = timeFull.substring(12, timeFull.length() - 1);
+                LocalDate date = LocalDate.parse(timeFull.substring(0, 11), formatter);
+                currentTask = new Deadline(taskName, date, time);
+            } else {
+                LocalDate date = LocalDate.parse(timeFull.substring(0, 11), formatter);
+                currentTask = new Deadline(taskName, date);
+            }
+            tasks.add(currentTask);
```
I like how there are blank lines between blocks of code serving different purposes, as it helps improve readability. Perhaps some comments can be given before each block to briefly explain what they do to make them more understandable!

nguyiyang commented on src/test/java/duke/command/ParserTest.java:
```diff
@@ -0,0 +1,27 @@
+
+import duke.exception.InvalidCommandException;
+
+public class ParserTest {
```
Maybe some descriptions for test methods can be provided. However, I think the JavaDoc method descriptions for the rest of your classes are clear and concise.

nguyiyang reviewed (commented):
Overall there wasn't much issues regarding coding standard violations from what I can see, except for naming of certain variables. More comments could also be provided for longer chunks of code to make them understood more easily. LGTM! :smiley:

zhoukerrr commented on src/main/java/duke/command/Ui.java:
```diff
@@ -0,0 +1,124 @@
+     *
+     * @param type The type of exception.
+     */
+    public void printException(String type) {
```
I like how you use the switch statements here. I can look into implementing this!

zhoukerrr commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+     * @param msg The error message of the exception.
+     */
+    public DukeException(String msg) {
+        super(msg);
```
Correct me if I am wrong. If the subclasses are just calling super, perhaps we do not really need this?

zhoukerrr reviewed (commented):
Nicely structured and very easy to follow! I get some inspiration. Thanks!

lwj1711 committed:
Complete Level-10

lwj1711 committed:
Merge branch 'branch-Level-10'

lwj1711 committed:
Improve exception handling

lwj1711 committed:
Update build.gradle file

lwj1711 committed:
Complete A-Assertions

lwj1711 committed:
Complete A-CodeQuality

lwj1711 committed:
Improve A-Code Quality

Current implementation of addTaskToList has many lines of code and might not adhere to SLAP.

Changes will ensure better readability for the Storage class.

Refractor code by extracting various methods.

Doing so adds three more methods to the current one, ensuring simplicity and better understandability.

lwj1711 committed:
Merge pull request #2 from lwj1711/branch-A-CodeQuality

Branch A-Code-Quality

lwj1711 committed:
Merge branch 'master' into branch-A-Assertions

lwj1711 committed:
Merge pull request #1 from lwj1711/branch-A-Assertions

Complete A-Assertions

lwj1711 committed:
Add missing semicolon

lwj1711 committed:
Complete A-CI

lwj1711 committed:
Complete C-Help

lwj1711 committed:
Merge branch 'branch-C-Help'

lwj1711 committed:
Add C-Help

Current implementation offers no guide for the user.

Users might be lost without being able to understand how to command the bot.

Add help command, so user can understand commands of the bot, while also providing examples.

This way, the bot will be more accessible and friendly for new users.

lwj1711 committed:
Improve UI greatly

lwj1711 committed:
Rename GUI to Duke

lwj1711 committed:
Fix minor styling

lwj1711 committed:
Add UI screenshot

lwj1711 committed:
Add javadoc for help function

lwj1711 committed:
Upload screenshots for different commands

lwj1711 committed:
Add User Guide

lwj1711 committed:
Update User Guide

lwj1711 committed:
Set theme jekyll-theme-modernist

lwj1711 committed:
Delete _config.yml

lwj1711 committed:
Modify files to solve file paths issue

lwj1711 committed:
Set theme jekyll-theme-slate

lwj1711 committed:
Add space for render of table in github pages
